### ISSUE TITLE
Enhancement Request: Skip all disk writes

### ISSUE THREAD

**Issue by agbernard** [classgraph/classgraph#236]
Love this library, but to use it I would need the ability to turn off all disk writes. Some of my users are not guaranteed to have write permissions.

I have an idea which I'll share as a PR shortly for how this might be achieved.

**Comment by agbernard** [comment]
> Interesting -- just out of curiosity, what sorts of runtime environments don't even have a temp dir?

Well they have temp directories, but all the apps may not have write permission to make use of it - an edge case for sure, but I have to account for it. Even if they have temp dir access, I can anticipate many of them not being comfortable with unpacking every nested jar on the classpath since they don't always have control of every jar that ends up on their classpath. It would be a much easier sell for me if I can just turn off disk writes altogether.

> I'm concerned that a no-op implementation will break behavior for a lot of your users.

I don't think it would be as bad as you imagine. Our current implementation uses some very basic home-grown scanning logic that doesn't come close to the set of things covered by `ClassGraph`. It only searched the default system classloader and didn't do any unpacking of nested jars. So even with disk writes off (and all the requisite consequences) we'd still be covering much more ground than we used to.

In other words, I'm fine with not getting whatever is not possible when unpacking is completely turned off so don't feel like you have maintain that behavior to satisfy my request. 

If you want to provide an alternate approach that maintains the unpacking without the disk writes I won't complain either :) Both of your alternatives seem viable, but if I had to pick I'd go with the "ram disk" approach - we're not starving for memory so I'd favor the more performant solution.

**Comment by lukehutch** [comment]
OK, I'll try to implement this as a 3-way option: extract to temp dir, extract to "ram disk", and skip classpath elements that would need to be extracted to be scanned.

This will take some time to get the classloading part right, and I'm traveling for the next 10 days or so, so I may not get to this until at least a couple of weeks from now.

**Comment by lukehutch** [comment]
Actually I just realized it's not possible to implement unzipping from RAM using the standard Java API, since `ZipFile` can only unzip from a `File`, not from a `byte[]` array. So I think something like the solution you already proposed is going to be the only option here.

It also just occurred to me though that the code as currently written should work fine if no temp dir is available, because classpath elements that throw an IOException should simply be skipped. Can you please test this on an environment without write access?

**Comment by lukehutch** [comment]
> If that's the case couldn't we just make the `NoopNestedJarHandler` mimic that behavior without any negative consequences?

There are no negative consequences of ClassGraph skipping a classpath element because an `IOException` was thrown while trying to read it. (This happens all the time for `FileNotFoundExceptions` when a classpath element doesn't exist, for example.) Did you actually try it?

> If the only 2 options are unpack or skip and it still works for those who don't manually turn on skipping while not having temp dir access, then I think everyone is happy.

There may still be a usecase for disabling inner jar scanning, for efficiency reasons more often than a user not having a temp dir -- but ClassGraph doesn't need `NoopNestedJarHandler` to accomplish that -- the unzipping routines there can just throw an exception if an unzip attempt is made during scanning with this option enabled.

**Comment by lukehutch** [comment]
I added `ClassGraph#disableNestedJarScanning()` to the master version, in case you or anyone else would like to disable the extraction of nested jars, whether or not there is a temp dir available.

The other case where unzipping may occur is in the case of trying to load classes from jarfiles for which the package root is not at the root of the jarfile, e.g. where the classes are stored at `META-INF/classes` or similar. In this case, during classloading, ClassGraph will try to extract `META-INF/classes` to a temp dir, and if it fails with `IOException`, classloading will fail. This seems like reasonable behavior, since there is no other reasonable way to load the classes, short of writing a custom classloader that can lead the classes directly from the zipfile, without extracting to a temporary file. (`URLClassLoader` cannot handle jarfiles for which the package root is not at the jarfile root.) This will happen only for Spring Boot jars or similar, when the scanning code is not itself running in the Spring Boot jar (so that it is not running inside the Spring Boot classloader), only when no disk write access is available, and only when ClassGraph is used for classloading. This is an obscure enough combination that it is probably not worth it to develop a whole new classloader for this case. If you do have customers that hit this problem, please let me know, and I'll work on it.

**Comment by agbernard** [comment]
> In this case, during classloading, ClassGraph will try to extract META-INF/classes to a temp dir, and if it fails with IOException, classloading will fail. This seems like reasonable behavior, since there is no other reasonable way to load the classes

Agreed. 

> I added `ClassGraph#disableNestedJarScanning()`

Works for me. I'm satisfied with these changes for this issue. Thanks!

**Comment by lukehutch** [comment]
Released in 4.0.7. Thanks for the request, and for your help getting a fix in place!

**Comment by lukehutch** [comment]
Just a heads-up, I figured out an easy way to implement the more complex classloader I described, at least for the case of non-empty package roots (eg. `META-INF/classes`), and this is deployed in 4.1.x, removing one of the key needs for creating temporary files. (Nested jars still need to be extracted to be scanned, so the `ClassGraph#disableNestedJarScanning()` method will still be useful for disabling this.)

**Comment by lukehutch** [comment]
A further update: I just committed to master some sweeping changes to nested jarfile handling (this will be released in 4.6.0 after further testing). This change introduces a custom zipfile central directory parser that is able to read jarfiles within jarfiles without extracting the inner (nested) jarfile to a temporary file on disk, as long as the inner jarfile is stored rather than deflated.

If the inner jarfile is deflated (which is rare -- zipping something twice doesn't usually reduce the size), it still needs to be inflated before it can be scanned. For small inner jarfiles (<32MB), the file is inflated to a `ByteBuffer` in RAM rather than to a temp file on disk. If the inner jarfile is 32MB or larger, the file is extracted to disk. If writing to disk fails for large jarfiles (e.g. there is no writeable temp dir, or the temp dir is full), then ClassGraph falls back again to extracting the jarfile to RAM instead of disk.

With this change in place, scanning is much faster for classpaths with a lot of nested jars, and temporary files are almost never created (and if they are created and that fails, ClassGraph can fail over to extracting to RAM).

This is a better fix than simply disabling nested jar scanning (unless you don't care to scan nested jars, which is possible, so I left that configuration option in place).
